### DISCUSSION TITLE
const class references

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/csharplang#2978] [Q&A]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/798116/const-class-references.html)._

---
I can do

static readonly TimeSpan MAXTIME = new TimeSpan(0, 0, 5);

so why can't I do 

const TimeSpan MAXTIME = new TimeSpan(0, 0, 5);

??

These are identical to a developer....

Then I can use MAXTIME as a default in optional parameters.

---
#### Visual Studio Feedback System on 10/30/2019, 00:05 AM: 

Thank you for taking the time to provide your suggestion. We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

**Comment by HaloFour** [comment]
Sounds like this belongs on [csharplang](https://github.com/dotnet/csharplang/issues).

`const` and `readonly` are far from the same thing, though.

**Comment by mseld2** [comment]
const value needs to be known at compile time. static readonly value can be determined at runtime.

**Comment by mv10-work** [comment]
Probably more importantly than compile-time vs. runtime is how it's _used_ ... `const` references are replaced by the literal value, whereas `static readonly` is retrieved by reference at runtime. This means using `const` in something like a library requires re-compiling everything that uses that `const`.

I see `public const` often enough that I'm starting to think it should either be disallowed completely or at least flagged with a warning. I ask questions about this behavior in interviews (only out of curiosity, I haven't actually interviewed anyone yet who has been aware of it).

**Comment by jrmoreno1** [comment]
@mseld2 : the value is known at compile time, for certain values of “known”. In this case it is completely known as a timespan is basically a wrapper around an Int64.  More generally , a way to embed non-primitive struts and classes in the resulting binary would be necessary, but it would also be useful for other things.

@mv10-work : how it is used is pretty much the runtime vs compile time issue.  And, yes if you change a const value you need to recompile all references to it for it to take effect. But that is in my experience both expected and desirable behavior.
